### PR TITLE
8302877: Speed up latin1 case conversions

### PR DESCRIPTION
**Pull request by eirbjo**
This PR suggests we speed up Character.toUpperCase and Character.toLowerCase for latin1 code points by applying the 'oldest ASCII trick in the book'.

This takes advantage of the fact that latin1 uppercase code points are always 0x20 lower than their lowercase (with the exception of two code points which uppercase out of latin1).

To verify the correctness of the new implementation, the test `Latin1CaseConversion` is added with an exhaustive verification of toUpperCase/toLowerCase for all latin1 code points.

The implementation needs to balance the performance of the various ranges in latin1. An effort has been made to favour operations on ASCII code points, without causing excessive regression for higher code points.

Performance is benchmarked for 7 chosen sample code points, each representing a range or a special-case.  Results in the first comment.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302877](https://bugs.openjdk.org/browse/JDK-8302877): Speed up latin1 case conversions

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [70c624d7](https://git.openjdk.org/jdk/pull/12623/files/70c624d7d95a431ff2eb48f9d54b3aacb8f2ce02)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12623/head:pull/12623` \
`$ git checkout pull/12623`

Update a local copy of the PR: \
`$ git checkout pull/12623` \
`$ git pull https://git.openjdk.org/jdk pull/12623/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12623`

View PR using the GUI difftool: \
`$ git pr show -t 12623`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12623.diff">https://git.openjdk.org/jdk/pull/12623.diff</a>

</details>

### CODE REVIEW

**Comment by eirbjo** [comment]
Benchmark results:

Baseline:

```
Benchmark                                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.Latin1CaseFolding.toLowerCase          low  avgt   15  1.267 ± 0.013  ns/op
Characters.Latin1CaseFolding.toLowerCase            A  avgt   15  1.657 ± 0.011  ns/op
Characters.Latin1CaseFolding.toLowerCase            a  avgt   15  1.258 ± 0.005  ns/op
Characters.Latin1CaseFolding.toLowerCase      A-grave  avgt   15  1.656 ± 0.011  ns/op
Characters.Latin1CaseFolding.toLowerCase      a-grave  avgt   15  1.270 ± 0.023  ns/op
Characters.Latin1CaseFolding.toLowerCase           mu  avgt   15  1.261 ± 0.006  ns/op
Characters.Latin1CaseFolding.toLowerCase           yD  avgt   15  1.260 ± 0.005  ns/op
Characters.Latin1CaseFolding.toUpperCase          low  avgt   15  1.284 ± 0.043  ns/op
Characters.Latin1CaseFolding.toUpperCase            A  avgt   15  1.264 ± 0.008  ns/op
Characters.Latin1CaseFolding.toUpperCase            a  avgt   15  1.818 ± 0.016  ns/op
Characters.Latin1CaseFolding.toUpperCase      A-grave  avgt   15  1.261 ± 0.015  ns/op
Characters.Latin1CaseFolding.toUpperCase      a-grave  avgt   15  1.822 ± 0.013  ns/op
Characters.Latin1CaseFolding.toUpperCase           mu  avgt   15  1.823 ± 0.006  ns/op
Characters.Latin1CaseFolding.toUpperCase           yD  avgt   15  1.822 ± 0.008  ns/op
```

PR:

```
Benchmark                                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.Latin1CaseFolding.toLowerCase          low  avgt   15  0.878 ± 0.005  ns/op
Characters.Latin1CaseFolding.toLowerCase            A  avgt   15  1.038 ± 0.009  ns/op
Characters.Latin1CaseFolding.toLowerCase            a  avgt   15  1.036 ± 0.007  ns/op
Characters.Latin1CaseFolding.toLowerCase      A-grave  avgt   15  1.357 ± 0.015  ns/op
Characters.Latin1CaseFolding.toLowerCase      a-grave  avgt   15  1.352 ± 0.003  ns/op
Characters.Latin1CaseFolding.toLowerCase           mu  avgt   15  1.273 ± 0.002  ns/op
Characters.Latin1CaseFolding.toLowerCase           yD  avgt   15  1.352 ± 0.004  ns/op
Characters.Latin1CaseFolding.toUpperCase          low  avgt   15  0.880 ± 0.013  ns/op
Characters.Latin1CaseFolding.toUpperCase            A  avgt   15  0.920 ± 0.071  ns/op
Characters.Latin1CaseFolding.toUpperCase            a  avgt   15  1.055 ± 0.013  ns/op
Characters.Latin1CaseFolding.toUpperCase      A-grave  avgt   15  1.394 ± 0.010  ns/op
Characters.Latin1CaseFolding.toUpperCase      a-grave  avgt   15  1.391 ± 0.009  ns/op
Characters.Latin1CaseFolding.toUpperCase           mu  avgt   15  1.597 ± 0.021  ns/op
Characters.Latin1CaseFolding.toUpperCase           yD  avgt   15  1.354 ± 0.003  ns/op
```

File: test/jdk/java/lang/Character/Latin1CaseFolding.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+/**
+ * @test
+ * @summary Provides exchaustive verification of Character.toUpperCase and Character.toLowerCase
```
Comment (naotoj) [inline test/jdk/java/lang/Character/Latin1CaseFolding.java:31]: typo: "exhaustive"?

**Review by naotoj — COMMENTED** [review]
Looks good to me. I'd rather not use "case folding", as to me it implies "normalizing" but this is simply lowercasing/uppercasing.

File: test/jdk/java/lang/Character/Latin1CaseFolding.java (hunk 1 above)
Comment (eirbjo) [reply test/jdk/java/lang/Character/Latin1CaseFolding.java]: I did an 'exchaustive' search for 'exchaustive' across the code base and found a two comments in `LocaleData` and `LocaleData.cldr` in jdk/test/jdk/sun/text/resources`.

Would you like me to update these as well while we're here, or should we avoid getting out the scope for this PR?

**Comment by eirbjo** [comment]
>  I'd rather not use "case folding", as to me it implies "normalizing" but this is simply lowercasing/uppercasing.

I guess I was looking for a generic term for uppercase/lowercase. I picked "case conversion" instead.

**Comment by eirbjo** [comment]
> Looks good to me.

Thanks for looking! Would you also be willing to file a JBS issue for this?

File: test/jdk/java/lang/Character/Latin1CaseFolding.java (hunk 1 above)
Comment (naotoj) [reply test/jdk/java/lang/Character/Latin1CaseFolding.java]: I'd appreciate it. I don't mind fixing it with this PR.

Comment (eirbjo) [reply test/jdk/java/lang/Character/Latin1CaseFolding.java]: Thanks Naoto, I have fixed the spelling in these two unrelated files.

**Comment by eirbjo** [comment]
A site note: Early and crude experiements using the Vector API indicate that the 'oldest ASCII trick in the book' vectorizes pretty well.

Here's a benchmark comparing the strings "helloworld" and "HelloWorld" 1024 times, followed by either 'A' or 'B' (to force a an expensive mismatch):

```
Benchmark                    (size)  Mode  Cnt     Score    Error  Units
EqualsIgnoreCase.scalar        1024  avgt   15  6225.624 ± 89.182  ns/op
EqualsIgnoreCase.vectorized    1024  avgt   15  1246.110 ± 14.767  ns/op
```

I have the feeling that most case-insensitive comparisons are pretty short, so not sure how useful this is IRL.

**Comment by eirbjo** [comment]
> I have the feeling that most case-insensitive comparisons are pretty short, so not sure how useful this is IRL.

There seems to be a win from strings of size 32 bytes upwards. (That's probably longer than most keys in TreeMaps using String.CASE_INSENSITIVE_ORDER, such as j.n.h.HttpHeaders)

```
Benchmark                    (size)  Mode  Cnt   Score   Error  Units
EqualsIgnoreCase.scalar          16  avgt    2  20.608          ns/op
EqualsIgnoreCase.scalar          32  avgt    2  36.510          ns/op
EqualsIgnoreCase.vectorized      16  avgt    2  18.601          ns/op
EqualsIgnoreCase.vectorized      32  avgt    2  12.795          ns/op
```

This is outside scope for this PR, I just wanted to leave a trace of this observation here for future record.

File: src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template [hunk 2]
```diff
@@ -135,30 +135,36 @@ class CharacterDataLatin1 extends CharacterData {
...
-        if (((val & $$maskLowerCase) != 0) && 
-                ((val & $$maskCaseOffset) != $$maskCaseOffset)) { 
-            int offset = val << $$shiftCaseOffsetSign >> ($$shiftCaseOffsetSign+$$shiftCaseOffset);
-            mapChar = ch + offset;
+        if (ch < 'A') { // Fast path for low code points
+            return ch;
         }
-        return mapChar;
+        int l = ch | 0x20; // Lowercase using 'oldest ASCII trick in the book'
+        if ( l <= 'z' // In range a-z
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template:142]:
```suggestion
        if (l <= 'z' // In range a-z
```

File: test/micro/org/openjdk/bench/java/lang/Characters.java [hunk 3]
```diff
@@ -80,4 +82,47 @@ public void codePointOf() {
...
 
     }
+
+    @BenchmarkMode(Mode.AverageTime)
+    @OutputTimeUnit(TimeUnit.NANOSECONDS)
+    @State(Scope.Thread)
+    @Warmup(iterations = 5, time = 1)
+    @Measurement(iterations = 5, time = 1)
+    @Fork(3)
+    public static class Latin1CaseConversions {
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/java/lang/Characters.java:92]: Not sure if qualifying this as "Latin1" is necessary, even though that's what you've focused on for this PR. We could easily add some codePoints outside of the latin1 range (now or later) without changing the test. 

While having a switch with some readable names is a nice touch I think we should additionally allow integer codePoint as-is to keep it in line with the outer class, e.g. `default -> Integer.parseInt(codePoint);`

**Review by cl4es — APPROVED** [review]
Looks good. Some nits inline

File: src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template (hunk 2 above)
Comment (eirbjo) [reply src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template]: Fixed! (My IDE does not highlight this code, making it a bit harder to spot mistakes like this)

File: test/micro/org/openjdk/bench/java/lang/Characters.java (hunk 3 above)
Comment (eirbjo) [reply test/micro/org/openjdk/bench/java/lang/Characters.java]: You are probably right that Latin1 is a bit narrow here, removing the prefix.

I added Integer.parseInt as the default, good idea!

**Comment by eirbjo** [comment]
Thanks for your review and JBS juggling, Claes! 

I'll wait for a final word from @naotoj before integrating.

**Review by naotoj — APPROVED** [review]
LGTM. Thanks for the fix!

**Comment by dholmes-ora** [comment]
The testcase change is failing to compile:
```
open/test/micro/org/openjdk/bench/java/lang/Characters.java:115: error: case, default, or '}' expected
```

**Comment by eirbjo** [comment]
> The testcase change is failing to compile:
> 
> ```
> open/test/micro/org/openjdk/bench/java/lang/Characters.java:115: error: case, default, or '}' expected
> ```

Darn, this was rather embarrassing!

Since this was just integrated, I guess I'll need to open a new PR for the fix?

**Comment by naotoj** [comment]
Yes please.

**Comment by eirbjo** [comment]
I created https://github.com/openjdk/jdk/pull/12701 for this.

Could someone please file a JBS?

**Comment by naotoj** [comment]
https://bugs.openjdk.org/browse/JDK-8303033

**Comment by eirbjo** [comment]
Thanks, #12701 is ready for review.

**Pull request closed without merging**